WoodStar/WoodStar/Login7Stream.cs

```diff
@@ -0,0 +1,186 @@
+            WriteOffsetLength(buffer, ref bufferOffset, ref dataOffset, Password);
+            WriteOffsetLength(buffer, ref bufferOffset, ref dataOffset, applicationName);
+            WriteOffsetLength(buffer, ref bufferOffset, ref dataOffset, serverName);
+            WriteOffsetLength(buffer, ref bufferOffset, ref dataOffset, extension);
+            WriteOffsetLength(buffer, ref bufferOffset, ref dataOffset, clientLibraryName);
+            WriteOffsetLength(buffer, ref bufferOffset, ref dataOffset, language);
+            WriteOffsetLength(buffer, ref bufferOffset, ref dataOffset, database);
+            buffer.WriteBytes(bufferOffset, new byte[6]); // Client Mac address
+            bufferOffset += 6;
+            WriteOffsetLength(buffer, ref bufferOffset, ref dataOffset, ""); // SSPI
+            WriteOffsetLength(buffer, ref bufferOffset, ref dataOffset, attachDbFileName);
+            WriteOffsetLength(buffer, ref bufferOffset, ref dataOffset, changePassword);
+            buffer.WriteIntLittleEndian(bufferOffset, 0); // SSPI Long
+            bufferOffset += 4;
+
+            WriteString(buffer, ref bufferOffset, hostName);
+            WriteString(buffer, ref bufferOffset, UserName);
+            WriteEncryptedPassword(buffer, ref bufferOffset, Password);
+            WriteString(buffer, ref bufferOffset, applicationName);
+            WriteString(buffer, ref bufferOffset, serverName);
+            WriteString(buffer, ref bufferOffset, clientLibraryName);
+            WriteString(buffer, ref bufferOffset, language);
+            WriteString(buffer, ref bufferOffset, database);
+            WriteString(buffer, ref bufferOffset, attachDbFileName);
+            WriteString(buffer, ref bufferOffset, changePassword);
+
+            static void WriteOffsetLength(byte[] buffer, ref int bufferOffset, ref ushort dataOffset, string value)
+            {
+                buffer.WriteUnsignedShortLittleEndian(bufferOffset, dataOffset);
+                buffer.WriteUnsignedShortLittleEndian(bufferOffset + 2, (ushort)value.Length);
+                dataOffset += (ushort)(value.Length * 2);
+                bufferOffset += 4;
+            }
+
+            static void WriteString(byte[] buffer, ref int bufferOffset, string value)
+            {
+                for (var i = 0; i < value.Length; i++)
+                {
+                    buffer.WriteByte(bufferOffset++, (byte)value[i]);
+                    buffer.WriteByte(bufferOffset++, 0);
+                }
+            }
+
+            static void WriteEncryptedPassword(byte[] buffer, ref int bufferOffset, string value)
+            {
+                for (var i = 0; i < value.Length; i++)
+                {
+                    var byteValue = (byte)value[i];
+                    var firstFourBits = byteValue & 0b11110000;
+                    var lastFourBits = byteValue & 0b00001111;
+                    var swappedValue = firstFourBits >> 4 + lastFourBits << 4;
```

Comment:
```suggestion
                    var swappedValue = (firstFourBits >> 4) + (lastFourBits << 4);
```

+ operator has higher priority than the `>>` or `<<`.